feat!: Introduce new `PriorityOrderedSet`

luanpotter opened the pull request:
This introduces a new type of ordered set that takes a mapper function instead of a comparing function, and thus is able to perform some optimizations much needed in Flame.

In order not to break backwards compatibility too much, the following approach is taken:

* `OrderedSet` becomes an "interface" (in Dart world that is an abstract class)
* We provide two implementations, `ComparingOrderedSet` is the old one, new one is `PriorityOrderedSet`
* You can use static methods on the `OrderedSet` interface to create any sets with helper functions
* `QueryableOrderedSet` takes in the underlying `OrderedSet` so it works with both versions

The new `PriorityOrderedSet` uses a splay tree map instead of splay tree set and thus caches the priorities on the map keys. That means you can safely update priorities or have dynamic priority functions and have consistent behaviour before rebalancing.

This is a much nicer underlying fix for https://github.com/flame-engine/flame/issues/3363 than how we did it on the Flame side.

spydon commented on lib/comparing_ordered_set.dart:
```diff
@@ -0,0 +1,163 @@
+/// This does not store the elements priorities, so it is susceptible to race
+/// conditions if priorities are changed while iterating.
+/// For a safer implementation, use [PriorityOrderedSet].
+class ComparingOrderedSet<E> extends OrderedSet<E> {
```
A bit of a strange name I think? 🤔

luanpotter replied:
couldn't think of better names

spydon commented on lib/comparing_ordered_set.dart:
```diff
@@ -0,0 +1,163 @@
+class ComparingOrderedSet<E> extends OrderedSet<E> {
+  // If the default implementation of `Set` changes from `LinkedHashSet` to
+  // something else that isn't ordered we'll have to change this to explicitly
+  // be `LinkedHashSet` (or some other data structure that preserves order).
```
Why not use `LinkedHashSet` directly?

luanpotter replied:
good call

spydon commented on lib/priority_ordered_set.dart:
```diff
@@ -0,0 +1,136 @@
+/// For an alternative implementation, use [ComparingOrderedSet].
+class PriorityOrderedSet<K extends Comparable<K>, E> extends OrderedSet<E> {
+  final K Function(E a) _priorityFunction;
+  late SplayTreeMap<K, Set<E>> _backingSet;
```
And here, should we explicitly use a sorted set implementation?

spydon reviewed (approved):
Great job! 

I'm not too fond of `comparing` as a name, maybe `MappedOrderedSet` or something instead?

luanpotter commented:
I also don't like the names, but MappedOrderedSet is essentially the other one (takes a mapping function)! the og one takes a comparator function instead

spydon commented:
> I also don't like the names, but MappedOrderedSet is essentially the other one (takes a mapping function)! the og one takes a comparator function instead

Ah right, of course, maybe the other one should be called `MappedOrderedSet` then since `priority` is an implementation detail for Flame?
For `Comparing`, it feels like it shouldn't be a verb/noun there, but an adjective?

luanpotter commented:
Yeah I agree with Mapping for the other one. One of the helper functions is already called that. We migh need to rework those names, which is fine.
For the OG, Comparing is an adjective no? The same way that Mapping is an adjective form of the verb map.

luanpotter commented:
But I am open to other names, maybe `ComparatorBased` would be more clear but it's clunky.

spydon commented:
> But I am open to other names, maybe `ComparatorBased` would be more clear but it's clunky.

I like that, go for it!

Oooor could we have `OrderedSet` not being abstract (and have that being the comparator based one) and `MappedOrderedSet` just extending it and overriding necessary parts? Or the other way around might be easier? :thinking:

luanpotter commented:
`MappedOrderedSet` is a completely different implementation, so what you could have is they both extend a `BaseOS` and then OG can be called `OS` - but I think the way I did is better because essentially we are encouraging people to only use the `OrderedSet` interface + the static methods to create them. People shouldn't even be looking at the specific implementations:


```
  OrderedSet set = OrderedSet.comparing/mapping/etc(parameters);
  // using the interface allows them to switch the impl at any time
```

luanpotter commented:
I'm happy with `ComparatorBase`, will update later today!

spydon commented:
> `MappedOrderedSet` is a completely different implementation, so what you could have is they both extend a `BaseOS` and then OG can be called `OS` - but I think the way I did is better because essentially we are encouraging people to only use the `OrderedSet` interface + the static methods to create them. People shouldn't even be looking at the specific implementations:
> 
> ```
>   OrderedSet set = OrderedSet.comparing/mapping/etc(parameters);
>   // using the interface allows them to switch the impl at any time
> ```

Alright, sounds good! I guess we should update some docs in the readme too.

luanpotter commented:
@spydon I spent some time thinking about and after much deliberation I reached these conclusions:
* the best names I think are `ComparingOrderedSet` and `MappingOrderedSet`, making it clear what each is based off and both consistently in template. I updated all references and names accordingly
* you cannot manually construct a LinkedHashSet in dart, as it is a lint issue. The literals are preferred, and when using the literals we have to use Set, or force a cast. I opted for leaving as is (we can always revist)
* I updated and revitalized the README file

spydon commented on README.md:
```diff
@@ -4,13 +4,13 @@
+It accepts either a comparator function that compares items for their priority or a mapper function that maps items to their priority.
 
-Unlike Dart's [SplayTreeSet](https://api.dartlang.org/stable/1.24.3/dart-collection/SplayTreeSet/SplayTreeSet.html), it allows for several different elements with the same priority to be added.
+Unlike Dart's [SplayTreeSet](https://api.dart.dev/dart-collection/SplayTreeSet-class.html) or [SplayTreeMap](https://api.dart.dev/dart-collection/SplayTreeMap-class.html) classes, it allows for several different elements with the same "priority" to be added.
```
We should line-break at 100, only links can overflow.

spydon commented on README.md:
```diff
@@ -20,28 +20,65 @@ A simple usage example:
+
+In order to create an `OrderedSet`, however, you can just use the static methods on the interface itself:
+
+* `comparing<E>([comparator])`: creates a `ComparingOrderedSet` with the given `Comparator`.
```
I think it is even clearer here to write `OrderedSet.`, even though it is mentioned in the sentence before.

spydon reviewed (approved):
Very nice! 😍
